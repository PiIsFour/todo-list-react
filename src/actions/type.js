const actionType = Object.freeze({
	addList: 'ADD_LIST',
	selectList: 'SELECT_LIST',
	deleteList: 'DELETE_LIST',
})

export default actionType
