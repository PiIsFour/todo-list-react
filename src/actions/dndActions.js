import ActionType from './type'

export const DragItemType = Object.freeze({
	todoItem: 'TODO_ITEM',
})

export const reorderTodo = (item, to) => ({
	type: ActionType.reorderTodo,
	item,
	to,
})
