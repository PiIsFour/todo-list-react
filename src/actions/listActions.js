import { uuid } from 'uuidv4'
import actionType from './type'

export const addListWithId = (name, id) => ({
	type: actionType.addList,
	name,
	id,
})

export const addListFactory = uuid => name => addListWithId(name, uuid())

export const addList = addListFactory(uuid)

export const selectList = (id, name) => ({
	type: actionType.selectList,
	id,
	name,
})
