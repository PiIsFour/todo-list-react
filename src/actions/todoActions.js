import * as R from 'ramda'
import { uuid } from 'uuidv4'
import actionType from './type'

export const addTodoWithId = (name, toList, id) => ({
	type: actionType.addTodo,
	name,
	id,
	toList: R.pick(['id', 'name'], toList),
})

export const addTodoFactory = uuid => (name, toList) => addTodoWithId(name, toList, uuid())

export const addTodo = addTodoFactory(uuid)

export const checkTodo = (id, name, done) => ({
	type: actionType.checkTodo,
	id,
	name,
	done,
})
