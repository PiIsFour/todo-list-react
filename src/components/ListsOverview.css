.lists-overview {
	grid-area: lists;
}

.lists-overview ul {
  font-size: 1.2rem;
  line-height: 1.7;
  list-style: circle;
  padding-left: 1.1em;
}

.lists-overview li {
  cursor: pointer;
}

.lists-overview li:hover {
  opacity: 0.7;
}

.lists-overview li.active {
  font-weight: 700;
}
