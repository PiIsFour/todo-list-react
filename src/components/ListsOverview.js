import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import * as R from 'ramda'
import './ListsOverview.css'

import AddInput from './AddInput'
import { addList, selectList } from '../actions/listActions'

export const ListsOverviewView = ({todoLists, selectedListId, dispatch}) => {
	const dispatchAddList = name => dispatch(addList(name))
	const dispatchSelectList = (id, name) => dispatch(selectList(id, name))
	const renderListItem = ({name, id}) => <li
		className={id === selectedListId ? 'active' : ''}
		key={id}
		onClick={() => dispatchSelectList(id, name)}
	>{name}</li>
	return <div className="lists-overview">
		<h2>My lists</h2>
		<ul>
			{todoLists.map(renderListItem)}
		</ul>
		<AddInput lable="new list name" onAdd={dispatchAddList} />
	</div>
}

const ListsOverview = () => {
	const todoLists = useSelector(R.prop('todoLists'))
	const selectedListId = useSelector(({selectedList}) => selectedList.id)
	const dispatch = useDispatch()
	return <ListsOverviewView todoLists={todoLists} selectedListId={selectedListId} dispatch={dispatch} />
}

export default ListsOverview
