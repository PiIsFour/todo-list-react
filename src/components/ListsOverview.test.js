import React from 'react'
import { shallow } from 'enzyme'

import { addListWithId, selectList } from '../actions/listActions'

import { ListsOverviewView, getTodoLists, getSelectedListId } from './ListsOverview'

describe('ListsOverview', () => {
	it('snapshot matches', () => {
		const todoLists = [
			{name: 'test name', id: 'some id'},
		]
		const wrapper = shallow(<ListsOverviewView todoLists={todoLists} dispatch={() => {}} />)
		expect(wrapper).toMatchSnapshot()
	})

	it('dispatches a addAndSelectList action', () => {
		const dispatch = jest.fn()
		const wrapper = shallow(<ListsOverviewView todoLists={[]} dispatch={dispatch} />)
		wrapper.find('AddInput').simulate('add', 'my new cool list')
		expect(dispatch).toBeCalledWith(expect.any(Function))
		dispatch.mock.calls[0][0](dispatch)
		expect(dispatch).toHaveBeenCalledWith(addListWithId('my new cool list', expect.toBeAUuid()))
		expect(dispatch).toHaveBeenLastCalledWith(selectList(dispatch.mock.calls[1][0].id, 'my new cool list'))
	})

	it('dispatches a selectList action', () => {
		const dispatch = jest.fn()
		const lists = [{id: 'test id', name: 'test list'}]
		const wrapper = shallow(<ListsOverviewView todoLists={lists} dispatch={dispatch} />)
		wrapper.find('ListOverviewItem').simulate('select', 'test id', 'test list')
		expect(dispatch).toHaveBeenCalledWith(selectList('test id', 'test list'))
	})

	it('selectedList list is active', () => {
		const dispatch = jest.fn()
		const lists = [
			{id: 'test id', name: 'test list'},
			{id: '42', name: 'Active list'},
		]
		const wrapper = shallow(<ListsOverviewView todoLists={lists} selectedListId="42" dispatch={dispatch} />)
		expect(wrapper.find('ul').childAt(0)).toHaveProp({isSelected: false})
		expect(wrapper.find('ul').childAt(1)).toHaveProp({isSelected: true})
	})

	it('getTodoLists returns the todo lists', () => {
		const state = {
			todoLists: [
				{id: '12'},
				{id: '42'},
				{id: '152'},
			],
			selectedList: {id: '42'},
		}
		expect(getTodoLists(state)).toBe(state.todoLists)
	})

	it('getSelectedListId returns the id of the selected list', () => {
		const state = {
			todoLists: [
				{id: '12'},
				{id: '42'},
				{id: '152'},
			],
			selectedList: {id: '42'},
		}
		expect(getSelectedListId(state)).toBe(state.selectedList.id)
	})
})
