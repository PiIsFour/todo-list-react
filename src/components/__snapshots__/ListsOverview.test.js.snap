// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsOverview snapshot matches 1`] = `
<div
  className="lists-overview"
>
  <h2>
    My lists
  </h2>
  <ul>
    <ListOverviewItem
      id="some id"
      isSelected={false}
      key="some id"
      name="test name"
      onSelect={[Function]}
    />
  </ul>
  <AddInput
    lable="new list name"
    onAdd={[Function]}
  />
</div>
`;
